modules/tetramesh/glsl/tetramesh_traversal.frag

```diff
@@ -0,0 +1,374 @@
+        tetraId = tetraFaceId / 4;
+        localFaceId = tetraFaceId % 4;
+        vertices = vertexIds[tetraId];
+
+        // query data of current tetrahedron
+        tetra = getTetra(tetraId);
+        mat4x3 faceNormal = getFaceNormals(tetra);
+
+        // intersect ray at current position with all tetra faces
+        vec4 vdir = vec4(dot(faceNormal[0], rayDirection),
+                         dot(faceNormal[1], rayDirection),
+                         dot(faceNormal[2], rayDirection),
+                         dot(faceNormal[3], rayDirection));
+        vec4 vt = vec4(dot(tetra.v[1] - pos, faceNormal[0]),
+                       dot(tetra.v[2] - pos, faceNormal[1]),
+                       dot(tetra.v[3] - pos, faceNormal[2]),
+                       dot(tetra.v[0] - pos, faceNormal[3])) / vdir;
+
+        // only consider intersections on the inside of the current triangle faces, that is t > 0.
+        // Also ignore intersections being parallel to a face
+        vt = mix(vt, vec4(invalidDepth), lessThan(vdir, vec4(0.0)));
+
+        // ignore self-intersection with current face ID, set distance to max
+        vt[localFaceId] = invalidDepth;
+
+        // closest intersection
+        // face ID of closest intersection
+        const int vface1 = vt.x < vt.y ? 0 : 1;
+        const int vface2 = vt.z < vt.w ? 2 : 3;
+        const int vface = vt[vface1] < vt[vface2] ? vface1 : vface2;        
+        const float tmin = vt[vface];
+
+        vec3 endPos = pos + rayDirection * tmin;
+
+        const float scalar = normalizeScalar(barycentricInterpolation(endPos, tetra));
+        const vec3 gradient = getTetraGradient(tetra);
+        const vec3 gradientWorld = geometry.dataToWorldNormalMatrix * gradient;
+        
+        const int numSteps = 100;
+        float tDelta = tmin / float(numSteps);
+        for (int i = 1; i <= numSteps; ++i) {
+            float s = mix(prevScalar, scalar, i / float(numSteps));
+
+            float tCurrent = tTotal + tDelta * i;
+#if defined(BACKGROUND_AVAILABLE)
+            if (!bgBlended && tBackground > tCurrent - tDelta && tBackground <= tCurrent) {
+                dvrColor += (1.0 - dvrColor.a) * background;
+                bgBlended = true;
+            }
+#endif // BACKGROUND_AVAILABLE
+
+            vec4 color = sampleTF(s);
+            if (color.a > 0) {
+                tFirstHit = tFirstHit == invalidDepth ? tCurrent : tFirstHit;
+
+#if defined(SHADING_ENABLED)
+                // perform illumination in world coordinates
+                vec3 worldPos = (geometry.dataToWorld * vec4(pos + rayDirection * tDelta * i, 1.0)).xyz;
+                color.rgb = APPLY_LIGHTING(lighting, color.rgb, color.rgb, vec3(1.0),
+                                           worldPos, gradientWorld, -rayDirWorld);
+#endif // SHADING_ENABLED
+
+                // volume integration along current segment
+                color.a = absorption(color.a, tDelta * opacityScaling * 0.1);
```

Comment:
This scaling seems a bit ad hoc. 
could use regular dvr compositing function here?

Reply:
`opacityScaling` is a basic scaling factor like `REF_SAMPLING_INTERVAL` in the exponent of the absorption term of the volume rendering integral. The reason being that the tetramesh raycasting traces rays in model/scene coordinates and not in texture coordinates [0,1]. Thus, the mesh will be rendered differently based on its actual scale. 
The variable name is borrowed from ParaView where it is used for the very same purpose, that is ray casting in scene coords.

Reply:
> could use regular dvr compositing function here?
No, since we need to adjust the scaling term. That is not exposed.

Reply:
It is exposed through `tIncr`?https://github.com/inviwo/inviwo/blob/5977b4617e26207e0e65541fc0f4fe918940f7d5/modules/opengl/glsl/utils/compositing.glsl#L43C40-L43C46

Reply:
You are correct, it is. Originally, I did not want to use that function due to the additional if statement
```glsl
if (tDepth == -1.0 && color.a > 0.0) tDepth = t;
```